XLS-0102-wasm-vm/README.md

```diff
@@ -0,0 +1,391 @@
+
+#### A.1.2. JIT (Just-In-Time)
+
+JIT compiles essentially as the code is run, or just before. This allows for additional caching and optimizations.
+
+However, there are a few issues with JIT. From [this Stellar blog](https://stellar.org/blog/developers/why-doesnt-soroban-use-a-jit), JIT-based VMs are also not as secure and are susceptible to “JIT Bombs.” It has longer start times and greater memory usage. Mac also does not support JIT. The use of JIT may also result in different gas costs on different machines depending on what is in the cache, which would be a consensus-breaking change. Therefore, **JIT cannot be used for our needs**.
+
+#### A.1.3. Interpreted
+
+Interpreted mode just runs the code, like a REPL.
+
+We decided on this mode because, well, the other two don't work for our needs, even though they're often more performant.
+
+### A.2. The Different WASM Implementations
+
+The 5 WASM VM implementations we investigated were:
+
+- [WasmEdge](https://wasmedge.org/) (used by Xahau)
+- [WasmTime](https://wasmtime.dev/)
+- [Wasmer](https://github.com/wasmerio/wasmer)
+- [Wasmi](https://github.com/wasmi-labs/wasmi) (used by Polkadot and Stellar’s Soroban)
+- [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime)
+
+#### A.2.1. Initial Investigation
+
+<img width="768" height="381" alt="image" src="https://github.com/user-attachments/assets/c3e4255c-f956-41f2-941f-90b643b3d567" />
+
+Based on these findings, we narrowed down the search to **WasmEdge** and **WAMR**, which we then did further performance testing and analysis on.
+
+#### A.2.2. Performance Analysis
+
+<img width="400" height="300" alt="image4" src="https://github.com/user-attachments/assets/7b82dfaa-e70a-4829-a987-20e6ae9ebb16" /><img width="400" height="300" alt="image2" src="https://github.com/user-attachments/assets/438255a5-cdae-4731-8f1a-0fe7218cb6ae" /><img width="400" height="300" alt="image1" src="https://github.com/user-attachments/assets/76fac8b0-63f6-43b1-b766-8a67799f5f59" />
+
+These graphs clearly show that WAMR is much more performant.
+
+## Appendix B: Other Memory Management Strategies Considered
+
+Options:
+
+1. Caller-Allocated: The contract developer (on the WASM side) allocates fixed size arrays for returning data. The user knows the pointer and the length.
+   - This is really easy to implement, but means that the WASM dev needs to do their own memory allocation.
+2. Host-Allocated: The host (rippled) allocates WASM memory in host functions and passes the pointer and the length to the WASM program.
+   - This is super easy to use for devs, as they don’t need to worry about allocation and all of that. However, more research is needed to determine how possible it is, because currently the only way that we know how to do this involves allocating a new page every time (to ensure the host isn’t overwriting addresses in use).
```

Comment:
```suggestion
   - This is super easy to use for devs, as they don’t need to worry about allocation. However, more research is needed to determine how possible it is, because currently the only way that we know how to do this involves allocating a new page every time (to ensure the host isn’t overwriting addresses in use).
```